src/gtest/test_joinsplit.cpp

```diff
@@ -0,0 +1,213 @@
+
+        // Perform the proof
+        proof = js->prove(
+            inputs,
+            outputs,
+            output_notes,
+            ciphertexts,
+            ephemeralKey,
+            pubKeyHash,
+            randomSeed,
+            hmacs,
+            nullifiers,
+            commitments,
+            vpub_old,
+            vpub_new,
+            rt
+        );
+    }
+
+    // Verify the transaction:
+    ASSERT_TRUE(js->verify(
+        proof,
+        pubKeyHash,
+        randomSeed,
+        hmacs,
+        nullifiers,
+        commitments,
+        vpub_old,
+        vpub_new,
+        rt
+    ));
+
+    // Recipient should decrypt
+    // Now the recipient should spend the money again
+    auto h_sig = js->h_sig(randomSeed, nullifiers, pubKeyHash);
+    ZCNoteDecryption decryptor(recipient_key.viewing_key());
+
+    auto note_pt = NotePlaintext::decrypt(
+        decryptor,
+        ciphertexts[0],
+        ephemeralKey,
+        h_sig,
+        0
+    );
+
+    auto decrypted_note = note_pt.note(recipient_addr);
+
+    ASSERT_TRUE(decrypted_note.value == 10);
+
+    // Insert the commitments from the last tx into the tree
+    tree.append(commitments[0]);
+    auto witness_recipient = tree.witness();
+    tree.append(commitments[1]);
+    witness_recipient.append(commitments[1]);
+    vpub_old = 0;
+    vpub_new = 1;
+    rt = tree.root();
+    pubKeyHash = random_uint256();
+
+    {
+        boost::array<JSInput, 2> inputs = {
+            JSInput(), // dummy input
+            JSInput(witness_recipient, decrypted_note, recipient_key)
+        };
+
+        SpendingKey second_recipient = SpendingKey::random();
+        PaymentAddress second_addr = second_recipient.address();
+
+        boost::array<JSOutput, 2> outputs = {
+            JSOutput(second_addr, 9),
+            JSOutput() // dummy output
+        };
+
+        boost::array<Note, 2> output_notes;
+
+        // Perform the proof
+        proof = js->prove(
+            inputs,
+            outputs,
+            output_notes,
+            ciphertexts,
+            ephemeralKey,
+            pubKeyHash,
+            randomSeed,
+            hmacs,
+            nullifiers,
+            commitments,
+            vpub_old,
+            vpub_new,
+            rt
+        );
+    }
+
+    // Verify the transaction:
+    ASSERT_TRUE(js->verify(
+        proof,
+        pubKeyHash,
+        randomSeed,
+        hmacs,
+        nullifiers,
+        commitments,
+        vpub_old,
+        vpub_new,
+        rt
+    ));
+}
+
+TEST(joinsplit, full_api_test)
+{
+    auto js = ZCJoinSplit::Generate();
+
+    test_full_api(js);
+
+    js->saveProvingKey("./zcashTest.pk");
+    js->saveVerifyingKey("./zcashTest.vk");
+
+    delete js;
+
+    js = ZCJoinSplit::Unopened();
+
+    js->preloadProvingKey("./zcashTest.pk");
+    js->loadProvingKey();
+    js->loadVerifyingKey("./zcashTest.vk");
+
+    test_full_api(js);
+
+    delete js;
+}
+
+TEST(joinsplit, note_plaintexts)
```

Comment:
At first I thought this was redundant with ``TEST(noteencryption, api)`` in ``test_noteencryption.cpp``, but it isn't because it's testing ``Note``'s API as well. Maybe add cross-referencing comments so that people looking at one of these tests can find the other.